### ISSUE TITLE
Out of Memory in BitMapFactory

### ISSUE THREAD

**Issue by monkichi99** [bumptech/glide#833]
I don't understand the following error.

The original image is 5000 x 5000 decode in ARGB_8888 which equate to 12.5MB, but I have configured glide to use the screen size of 1920 x 1200 which should roughly equate to 1.2MB. But why is it trying to allocate BitMap of the original size and not the scaled down size?

Here is my glide configuration

```java
GlideLoadConfig config = new GlideLoadConfig.Builder()
    .fitCenter()
    .withDecodeFormat(DecodeFormat.PREFER_ARGB_8888)
    .build();
```
Extended ImageLoader code
```java
@Override
public Target getLoadTarget(@NonNull ImageView imageView) {
    return new TransitionImageViewTarget(imageView.getContext(), imageView);
}

@Override
protected void appendAdditionalParameters(int index, @NonNull Item item,
                                          @NonNull GenericRequestBuilder<Item, ?, ?, ?> builder) {
    super.appendAdditionalParameters(index, item, builder);
    //calculated screen size of 1920 x 1200
    builder.override(size.getWidth(), size.getHeight());
    builder.priority(Priority.HIGH);
    builder.signature(new StringSignature(item.getObjectId().toString()));
}
```
```java
public class TransitionImageViewTarget extends BitmapImageViewTarget {

    private final Context context;

    private ImageView view;

    public TransitionImageViewTarget(Context context, ImageView view) {
        super(view);
        this.context = context;
        this.view = view;
    }

    @Override
    public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
        Drawable current = view.getDrawable();
        if (current != null) {
            GlideTransitionDrawable transitionDrawable = new GlideTransitionDrawable(context, resource, current);
            view.setImageDrawable(transitionDrawable);
        } else {
            super.onResourceReady(resource, glideAnimation);
        }
    }
}
```

Loading the image into the `ImageView`:
```java
BitmapRequestBuilder<T, Bitmap> builder = buildBaseRequest(item, getViewActivity(imageView));
builder.placeholder(placeholderDrawable);
       builder.listener(new RequestListenerTuple<>(loadListener, requestListener));
appendGlideConfigParameters(builder, glideLoadConfig);
appendAdditionalParameters(index, item, builder);

if (requestListener != null) {
    requestListener.onLoadStarted();
}

builder.into(getLoadTarget(imageView));
```

```
D/dalvikvm(26154): GC_FOR_ALLOC freed 7935K, 31% free 36946K/53063K, paused 74ms
I/dalvikvm-heap(26154): Forcing collection of SoftReferences for 25000016-byte allocation
D/dalvikvm(26154): GC_BEFORE_OOM freed 12K, 31% free 36933K/53063K, paused 77ms
E/dalvikvm-heap(26154): Out of memory on a 25000016-byte allocation.
I/dalvikvm(26154): "fifo-pool-thread-0" prio=5 tid=71 RUNNABLE
I/dalvikvm(26154):   | group="main" sCount=0 dsCount=0 obj=0x41ee51e0 self=0x12b5838
I/dalvikvm(26154):   | sysTid=26404 nice=10 sched=0/0 cgrp=bg_non_interactive handle=19541776
I/dalvikvm(26154):   | schedstat=( 0 0 0 ) utm=1923 stm=377 core=0
I/dalvikvm(26154):   at android.graphics.BitmapFactory.nativeDecodeStream(Native Method)
I/dalvikvm(26154):   at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:493)
I/dalvikvm(26154):   at com.bumptech.glide.load.resource.bitmap.Downsampler.decodeStream(Downsampler.java:329)
I/dalvikvm(26154):   at com.bumptech.glide.load.resource.bitmap.Downsampler.downsampleWithSize(Downsampler.java:220)
I/dalvikvm(26154):   at com.bumptech.glide.load.resource.bitmap.Downsampler.decode(Downsampler.java:153)
I/dalvikvm(26154):   at com.bumptech.glide.load.resource.bitmap.StreamBitmapDecoder.decode(StreamBitmapDecoder.java:50)
I/dalvikvm(26154):   at com.bumptech.glide.load.resource.bitmap.StreamBitmapDecoder.decode(StreamBitmapDecoder.java:19)
I/dalvikvm(26154):   at com.bumptech.glide.load.resource.bitmap.ImageVideoBitmapDecoder.decode(ImageVideoBitmapDecoder.java:39)
I/dalvikvm(26154):   at com.bumptech.glide.load.resource.bitmap.ImageVideoBitmapDecoder.decode(ImageVideoBitmapDecoder.java:20)
I/dalvikvm(26154):   at com.bumptech.glide.load.engine.DecodeJob.decodeFromSourceData(DecodeJob.java:190)
I/dalvikvm(26154):   at com.bumptech.glide.load.engine.DecodeJob.decodeSource(DecodeJob.java:177)
I/dalvikvm(26154):   at com.bumptech.glide.load.engine.DecodeJob.decodeFromSource(DecodeJob.java:128)
I/dalvikvm(26154):   at com.bumptech.glide.load.engine.EngineRunnable.decodeFromSource(EngineRunnable.java:122)
I/dalvikvm(26154):   at com.bumptech.glide.load.engine.EngineRunnable.decode(EngineRunnable.java:101)
I/dalvikvm(26154):   at com.bumptech.glide.load.engine.EngineRunnable.run(EngineRunnable.java:58)
I/dalvikvm(26154):   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:442)
I/dalvikvm(26154):   at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
I/dalvikvm(26154):   at java.util.concurrent.FutureTask.run(FutureTask.java:137)
I/dalvikvm(26154):   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
I/dalvikvm(26154):   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
I/dalvikvm(26154):   at java.lang.Thread.run(Thread.java:856)
```

**Comment by TWiStErRob** [comment]
I don't know how you calculated, here's my calculation:
```
8888 takes 8+8+8+8=32 bits = 4 bytes per pixel
5000*5000*4=100MB
1920*1080*4=8.2MB
2500*2500*4=25MB
```
So this 25MB allocation may be something else than 5000-big image. Can you please validate if your `size` object indeed contains what you expect by printing it to the log or debug?

Your custom builder is rather complicated with a lot of moving parts, some omitted. Can you please create a more minimalist repro case where all the builder code is in one place (one statement)? Also try to remove irrelevant parts to this issue, like `priority` and `signature`.

You could also put a breakpoint on `GenericRequest.onSizeReady` to see if the size arguments are what you're expecting.

There's a potential lurking bug at `view.getDrawable()`, read up on this [here](https://github.com/bumptech/glide/issues/527#issuecomment-119232936).

(Note: I added some formatting to your post to make it easier to comprehend, check them out by editing (upper right corner of the comment) and also read: https://guides.github.com/features/mastering-markdown/)

**Comment by TWiStErRob** [comment]
From what I understood from above though, it looks ok, and shouldn't load a too big image. There's one possibility that I just realized: when you're downsampling a huge image, downsampling can only happen in 2s powers: `Bitmap.Options.inSampleSize`. The default downsampling is `Downsampler.AT_LEAST` which means the bitmap loaded by `BitmapFactory` will be at least the requested size `1920x1080`, so in the case of your 5000px big image that one of 5000x5000, 2500x2500, 1250x1250, 625x625, ... will be loaded. From these `2500x2500` fits the "at least" criteria (`Math.min(5000/1920, 5000/1080) == Math.min(2, 4)`), so that's what Glide will aim for. This is then later reduced to the exact size of `1920x1080` via the `FitCenter` transformation. This "at least" sizing is there to prevent loading an `1250x1250` image and then later upscale to `1920x1080` which would reduce quailty a lot.

So to sum up: to load a `5000x5000` image with `override(1920, 1080)` you need free memory of `25MB + 8.2MB` for a moment, and then the 25MB image is discarded and the 8.2MB `Bitmap` is delivered.

It looks like you're already using `asBitmap` somewhere hidden, so you can call `atMost()` on the builder, but you'll lose quailty. Another option is to use `overrride` with a smaller expected image size.

**Comment by faifai21** [comment]
I've got the same issue. I have a bunch of images with very high resolutions (4000x3000). I have some conditional logic to choose whether to load the image from the server or locally (if it is available). When loading from the server, I can ask the server to downscale the image to a much more manageable size, which would be a function of the screen size. If the image is not available on the server, or if we have no network to work with, we try to load the image locally.

When loading local images, we still calculate the image width as a function of the screen size. Usually, this ends up in a call to `.override(width, Target.SIZE_ORIGINAL).fitCenter()` where `width = screenWidth`. The `Target.SIZE_ORIGINAL` is there to (hopefully) preserve the aspect ratio without needing to do the calculation ourselves.

On a Samsung SPH-M830 running 4.1.2, I always run OOM when trying to load a scaled down version of the local image. I suspect this is due to what @TWiStErRob said:

> to load a 5000x5000 image with override(1920, 1080) you need free memory of 25MB + 8.2MB for a moment, and then the 25MB image is discarded and the 8.2MB Bitmap is delivered.

Is there a way to avoid allocation that first bit of memory? On older devices, that first allocation will shut take up all the memory on the device.

**Comment by TWiStErRob** [comment]
Note that the 25MB is to load the `2500x2500` image and which is already a huge saving compared to the 100MB that would be required for the full image... this saving gets much better if you load small thumbnails, but we can't really work around the BitmapFactory limitation of 2s power downsampling.

One way to avoid that would be to skip the fitCenter operation and choose a downsampling size that undershoots the value, so the quality may be tid bit worse, but you'll only need about 6MB in the above case to load a single Bitmap of size 1250x1250, code (untested):

```java
load(...)
.asBitmap()
.atMost()
.override(your, size)
.dontTransform()
```
and a `scaleType="fitCenter"` in the XML.

The problem with `override(one argument SIZE_ORIGINAL)` is that as far as I remember the default downsampling operation is `atLeast` which may force Glide to load the full-sized image and then transform it further down in memory. The reason for this default is that because there's a transformation after load there should be more pixels available to be able to get a better quality image: if the default was `atMost` then the size would be less than the target size and the transformations would either upscale or don't do anything. Note that it is safe to pass `.override(width, width)` the aspect will be always preserved while decoding or using built-in transformations (centerCrop/fitCenter).

Much of this is speculation based on some vague memories, try to add breakpoints at these points in the code to verify for your particular image and code (check the inputs to the method calls at the assignments and the value they return to see what happened inside):
 * `GenericRequest.onSizeReady` at `loadStatus = engine.load`
 * `Downsampler.decode` at `downsampled = downsampleWithSize`
 * `BitmapTransformation.transform` at `transformed = transform`
 * `ImageViewTarget.onResourceReady` the parameter `resource`  
   (see also `super.view.getWidth()` to see how big it will be displayed

**Comment by faifai21** [comment]
Awesome! Your suggestion to remove `fitCenter()` (we were already applying the scale type on the image view) and replace `SIZE_ORIGINAL` with the width worked wonders. I'm now able to resize and load those images using Glide without running OOM.

Thanks @TWiStErRob!!

**Comment by TWiStErRob** [comment]
Good to hear that, I hope by "remove fitCenter()" you mean you also added `dontTransform`, because otherwise the default `scaleType` on `ImageView` is `fitCenter` (even if not in XML) and that will be picked up by `into(ImageView)` if there's no other transformation (including "no transformation") is set.

**Comment by faifai21** [comment]
> I hope by "remove fitCenter()" you mean you also added dontTransform

Yep. Thanks again!
